styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java

```diff
@@ -252,21 +253,31 @@ private StatusType eventInjectorHelper(Event event) {
       return Response.forStatus(BAD_REQUEST.withReasonPhrase(e.getMessage()));
     }
 
-    // Verifying active
-    if (stateManager.isActiveWorkflowInstance(workflowInstance)) {
-      return Response.forStatus(
-          BAD_REQUEST.withReasonPhrase("The specified instance is already "
-                                       + "active in the scheduler"));
-    }
-
     // Verifying future
     if (instant.isAfter(time.get())) {
       return Response.forStatus(BAD_REQUEST.withReasonPhrase(
           "Cannot trigger an instance of the future"));
     }
 
     final String triggerId = randomGenerator.generateUniqueId(AD_HOC_CLI_TRIGGER_PREFIX);
-    triggerListener.event(workflow, Trigger.adhoc(triggerId), instant);
+    final CompletionStage<Void> triggered = triggerListener.event(workflow, Trigger.adhoc(triggerId), instant);
+
+    // TODO: return future instead of blocking
+    try {
+      triggered.toCompletableFuture().get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      final Throwable cause = e.getCause();
+      if (cause instanceof IllegalStateException
+          || cause instanceof IllegalArgumentException) {
+        // TODO: propagate error information using a more specific exception type
+        return Response.forStatus(BAD_REQUEST.withReasonPhrase(cause.getMessage()));
```

Comment:
Should use `CONFLICT` for `IllegalStateException`? It was not a malformed request.